### PR TITLE
Support Installation Specification

### LINKED ISSUE #1177: Multiple installations with the same name?
I may have missed an issue in the backlog related to this, but I'm wondering if we need additional logic in Porter to allow for (or, conversely, enforce the impossibility) of managing multiple installations with the same bundle name.  (Recall that bundle name represents the default installation name, as well as the default name for the credential and parameter set files.)

Today, it seems that if I had installed a bundle named `mybundle`, with correspondingly named credential and parameter sets, Porter doesn't intervene if I go to install a different bundle with the same name (`mybundle`): the installations records, credential and parameter sets all appear to overwrite the former installation's assets.

Thoughts on what Porter should enforce and/or improve upon in this scenario?

### LINKED ISSUE #1683: Support namespaces and labels
In preparation of implementing advanced dependency scenarios from [PEP #003 - Dependency Namespaces and Labels](https://github.com/getporter/proposals/pull/8), let's add the concept of namespaces and labels.

### PR DESCRIPTION
**Pull request by carolynvs**
# What does this change
* Move all documents and data access code into Porter
* Funnel most access to cnab-go/claims through porter's own pkg/claims (soon to be renamed to pkg/installations). I was seeing bugs and compilation errors because it's so easy to confuse the two or accidentally use the struct from the wrong package. This makes sure that all cnab installation data stuff goes through Porter.
* Base storage plugin protocol on mongodb
* Remove BackingStore and all autoconnect logic. We now connect to the storage plugin just in time, connecting as needed and holding the connection until porter exits. This means that tests that access storage must call `Teardown()` to release the connection.
* Change filesystem storage plugin to run an instance of mongo and write the db files to PORTER_HOME.
* Add namespace to all documents
* Support namespaces, e.g. `porter list -n dev`, PORTER_NAMESPACE and namespace config
* Run internal plugins _inside_ the porter process, instead of execing back into itself. Yes this means that internal plugins are treated a bit differently. The protocol is still the same though and it greatly improves debugging.

TODO FOR THIS PR
- [x] Rename "filesystem" plugin to "docker-mongodb" (in this PR)
- [ ] Document the mongodb-docker plugin (In this PR)

TODO AFTER THIS PR
- [ ] Switch from using docker CLI directly, to using the docker library (in follow-up PR)
- [ ] Add labels to all documents (in follow-up PR)
- [ ] Add --all-namespaces, which defaults --namespace to "*" appropriately (in folow-up PR)

# What issue does it fix
Fixes #1177
First part of #1683

# Notes for the reviewer
Sorry it's giant, I needed to make sure the entire thing worked end-to-end. It may work better to do the review in a couple screen sharing sessions.

# Checklist
- [ ] Unit Tests
- [ ] Documentation
- [ ] Schema (porter.yaml)

### CODE REVIEW

File: pkg/porter/porter.go [hunk 1]
```diff
@@ -31,45 +36,86 @@ type Porter struct {
...
+	// Start up our plugins
+	// TODO(carolynvs): have each plugin store (storage and secrets) connect on demand by wrapping the underlying store like the manager does.
+	err := p.Secrets.Connect()
+	if err != nil {
+		return err
+	}
+	return p.Storage.Connect()
+}
+
+// Close releases resourced used by Porter before terminating the application.
```
Comment (vdice) [inline pkg/porter/porter.go:100]:
```suggestion
// Close releases resources used by Porter before terminating the application.
```

File: pkg/secrets/pluginstore/doc.go [hunk 2]
```diff
@@ -1,3 +1,3 @@
 // Package pluginstore is an internal Porter package that implements the
-// secret.Store interface via plugins.
+// secret.CredentialStore interface via plugins.
```
Comment (vdice) [inline pkg/secrets/pluginstore/doc.go:2]: Was this an unintended update?

File: pkg/secrets/pluginstore/store.go [hunk 3]
```diff
@@ -39,38 +43,54 @@ func NewSecretsPluginConfig() pluggable.PluginTypeConfig {
...
 		return err
 	}
 	s.cleanup = cleanup
 
-	store, ok := raw.(cnabsecrets.Store)
+	store, ok := raw.(plugins.SecretsProtocol)
 	if !ok {
 		cleanup()
-		return errors.Errorf("the interface exposed by the %s plugin was not secrets.Store", l.SelectedPluginKey)
+		return errors.Errorf("the interface exposed by the %s plugin was not secrets.CredentialStore", l.SelectedPluginKey)
```
Comment (vdice) [inline pkg/secrets/pluginstore/store.go:83]:
```suggestion
		return errors.Errorf("the interface exposed by the %s plugin was not plugins.SecretsProtocol", l.SelectedPluginKey)
```

File: pkg/storage/migrations/helpers.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+		testPlugin: testPlugin,
+		Manager:    NewManager(c.Config, storage.NewPluginAdapter(testPlugin)),
+	}
+}
+
+func (m *TestManager) Teardown() error {
+	return m.testPlugin.Teardown()
+}
+
+// UnsetSchema allows tests to pre-emptively set the schema document.
```
Comment (vdice) [inline pkg/storage/migrations/helpers.go:27]:
```suggestion
// SetSchema allows tests to pre-emptively set the schema document.
```

File: pkg/storage/plugin_adapter.go [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+	"get.porter.sh/porter/pkg/storage/plugins"
+	"github.com/globalsign/mgo/bson"
+	"github.com/pkg/errors"
+)
+
+var _ Store = PluginAdapter{}
+
+// PluginAdapter converts between the low-level plugin.StorageProtocol which
+// operators on bson documents, and the document types stored by Porter which are
```
Comment (vdice) [inline pkg/storage/plugin_adapter.go:16]:
```suggestion
// operates on bson documents, and the document types stored by Porter which are
```

File: pkg/storage/plugins/mongodb/mongodb.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+	c := s.getCollection(session, opts.Collection)
+	return c.Find(opts.Filter).Count()
+}
+
+func (s *Store) Find(opts plugins.FindOptions) ([]bson.Raw, error) {
+	session, err := s.copySession()
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
```
Comment (vdice) [inline pkg/storage/plugins/mongodb/mongodb.go:124]: nit: just noticed this could probably be simplified to `defer session.Close()` like the others?

File: pkg/storage/pluginstore/store.go [hunk 7]
```diff
@@ -37,39 +44,113 @@ func NewStoragePluginConfig() pluggable.PluginTypeConfig {
...
 		return errors.Wrapf(err, "could not load %s plugin", pluginType.Interface)
 	}
 	s.cleanup = cleanup
 
-	store, ok := raw.(crud.Store)
+	store, ok := raw.(plugins.StorageProtocol)
 	if !ok {
 		cleanup()
-		return errors.Errorf("the interface exposed by the %s plugin was not crud.Store", l.SelectedPluginKey)
+		return errors.Errorf("the interface exposed by the %s plugin was not crudstore.StorageProtocol", l.SelectedPluginKey)
```
Comment (vdice) [inline pkg/storage/pluginstore/store.go:79]:
```suggestion
		return errors.Errorf("the interface exposed by the %s plugin was not plugins.StorageProtocol", l.SelectedPluginKey)
```

**Review by vdice — COMMENTED** [review]
👏 Impressive work, as always!  These changes set Porter up to be far more performant and efficient.  Majority of comments are minor and usually just typo-related.

It looks like there may be one more TODO in the description above, around documenting the mongodb-docker plugin -- does that still need to be added?

File: pkg/claims/installation.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+	i.Status.RunID = run.ID
+	i.Status.Action = run.Action
+	i.Status.ResultID = result.ID
+	i.Status.ResultStatus = result.Status
+
+	if !i.Status.InstallationCompleted && run.Action == cnab.ActionInstall && result.Status == cnab.StatusSucceeded {
+		i.Status.InstallationCompleted = true
+	}
+
+	// TODO(carolynvs): set this before running the bundle, it should trigger the execution
```
[the remark this answers is not in the document]
Comment (carolynvs) [reply pkg/claims/installation.go:115]: That is part of #1704

**Comment by carolynvs** [comment]
Oh! Yes you are right, I wanted to document the plugin first before merging this PR. 👍

File: pkg/claims/output.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	ResultID     string `json:"resultId"`
+	Value        []byte `json:"value"`
+}
+
+func (o Output) DefaultDocumentFilter() interface{} {
+	return map[string]interface{}{"resultId": o.ResultID, "name": o.Name}
+}
+
+func (o Output) ToCNAB() cnab.Output {
+	return cnab.Output{
```
[the remark this answers is not in the document]
Comment (carolynvs) [reply pkg/claims/output.go:28]: Oops, it's because that function isn't needed. I'll remove it and the others. Right now the only thing that needs it is Run -> Claim, and perhaps later Installation as well.

File: pkg/claims/provider.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+	ListInstallations(namespace string) ([]Installation, error)
+
+	ListRuns(namespace string, installation string) ([]Run, error)
+	ListResults(runID string) ([]Result, error)
+	ListOutputs(resultID string) ([]Output, error)
+
+	GetRun(id string) (Run, error)
+	GetResult(id string) (Result, error)
+
+	/*
```
[the remark this answers is not in the document]
Comment (carolynvs) [reply pkg/claims/provider.go:26]: Oops yeah that's from when I wasn't sure what functions I would need. Removed!

File: pkg/claims/run.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+	"github.com/cnabio/cnab-go/bundle"
+	"github.com/cnabio/cnab-go/schema"
+)
+
+var _ storage.Document = Run{}
+
+// Run represents the execution of an installation's bundle.
+type Run struct {
+	// SchemaVersion is the version of the claim schema.
+	SchemaVersion schema.Version `json:"schemaVersion"`
```
[the remark this answers is not in the document]
Comment (carolynvs) [reply pkg/claims/run.go:17]: Thanks, yeah I missed that when I changed my mind about how to handle encoding.

File: pkg/config/config.go [hunk 12]
```diff
@@ -79,7 +79,18 @@ func (c *Config) LoadData() error {
 		c.DataLoader = LoadFromEnvironment()
 	}
 
-	return c.DataLoader(c)
+	if err := c.DataLoader(c); err != nil {
+		return err
+	}
+
+	if c.Debug {
+		ns := "(global)"
```
[the remark this answers is not in the document]
Comment (carolynvs) [reply pkg/config/config.go:87]: I was trying to be clear in the debug statement what namespace we were using even when the string was empty (since printing the empty string looked wrong). I've changed it to print `''` to be consistent.

File: pkg/parameters/parameterset.go [hunk 13]
```diff
@@ -19,26 +20,36 @@ const (
 	CNABSpecVersion string = "cnab-parametersets-" + string(DefaultSchemaVersion)
```
[the remark this answers is not in the document]
Comment (carolynvs) [reply pkg/parameters/parameterset.go:20]: It actually wasn't necessary at all, so I have removed it. Parameter Sets are never passed back _into_ cnab-go, so the value isn't needed.

File: pkg/secrets/pluginstore/doc.go (hunk 2 above)
Comment (carolynvs) [reply pkg/secrets/pluginstore/doc.go:2]: When copy/paste goes wrong

**Comment by carolynvs** [comment]
I've fixed all the feedback I think. I'm going to leave the documentation for Monday and peace out for the weekend. Thanks for the review!

**Comment by carolynvs** [comment]
https://deploy-preview-1684--porter.netlify.app/plugins/
https://deploy-preview-1684--porter.netlify.app/plugins/mongodb/
https://deploy-preview-1684--porter.netlify.app/plugins/mongodb-docker/

File: docs/content/plugins/mongodb-docker.md [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+The plugin runs a MongoDB server in a container, storing its data on a separate
+volume. The container is named `porter-mongodb-docker-plugin` and the volume is
+named `porter-mongodb-docker-plugin-data`. The plugin leaves the container
+running in-between Porter commands for performance reasons. It is safe to stop
+or remove the container. Removing the volume will result in data loss.
+
+## Plugin Configuration
+
+To use the mongodb plugin, add the following config to porter's config file
+(default location: `~/.porter/config.toml`). Replace `conn_str` with the
```
Comment (vdice) [inline docs/content/plugins/mongodb-docker.md]: `Replace conn_str with the connection string for your MongoDB server.` should be removed, right?

**Review by vdice — APPROVED** [review]
One minor docs Q, otherwise LGTM!

File: docs/content/plugins/mongodb-docker.md (hunk 14 above)
Comment (carolynvs) [reply docs/content/plugins/mongodb-docker.md]: oops! thanks
